### ISSUE TITLE
`v9.x` FAQ

### ISSUE THREAD

**Issue by lukekarrys** [npm/cli#5844]
### Breaking Changes

- For a full list of the breaking changes, see [this GitHub Changelog post](https://github.blog/changelog/2022-10-24-npm-v9-0-0-released/) or the initial [`npm@9.0.0`](https://github.com/npm/cli/releases/tag/v9.0.0) release

### Installing symlinked local dependencies

`npm@9` changed the default value of `install-links` from `false` to `true`.

This means that npm will now attempt to install directories by default instead of symlinking them.

Note that workspaces will always be symlinked, so the latest workspaces changes will always be reflected in your package.

**How to turn it off**

You can run the following to set the previous behavior on a per-project basis:

```
npm config set install-links=false --location project
```

### Compatibility with `npm@6`

`npm@9` changed the `lockfile-version` for new projects from `2` to `3`.

If you are creating a **new project** and want to maintain compatibility with `npm@6` you will need to downgrade your `lockfile-version` to `2`. Existing projects will not have their `lockfile-version` number changed.

To set a project specific `lockfile-version` run:

```
npm config set lockfile-version=2 --location project
```

### `ENOENT` `package.json` Errors

> **Note**:
> If you get this error, try running the command with `--install-links=false`

`npm@9` has changed the default of the `install-links` flag to `true`. This means that when running `npm install` with a directory as the argument (such as `npm install ../path/to/local-package`), `npm` will attempt to install that directory into the `node_modules/` directory just like any other package, instead of creating a symlink like it would previously.

We are tracking a fix (ref: https://github.com/npm/cli/pull/5820) for a known issue where transitive local directory dependencies can cause this error.

**Workaround**

To workaround this issue, run either of the following:
- `npm install --install-links=false` to run a single install that will create symlinks
- `npm config set install-links=false --location project` to set this as a default for your project in its local `.npmrc file`

## Other Issues

If you find other issues with `v9`, use the [Bug v9 Issue Template](https://github.com/npm/cli/issues/new?assignees=&labels=Bug%2CNeeds+Triage%2CRelease+9.x&template=bug_9.yml&title=%5BBUG%5D+%3Ctitle%3E) to open a new issue.

**Comment by jeffrson** [comment]
npm@9 was announced by my current npm@8 - and it should be GA now - npm@latest is 9.1.2.

Unfortunately something goes wrong with auth-type "web":
```
npm info using npm@9.1.2
npm info using node@v18.12.1
npm verb title npm adduser
npm verb argv "adduser" "--loglevel" "verbose" "--auth-type" "web" "--registry" "http://localhost:4873"
npm verb logfile logs-max:10 dir:...
npm verb logfile ...
npm notice Log in on http://localhost:4873/
npm verb web adduser before first POST
npm http fetch POST 404 http://localhost:4873/-/v1/login 33ms
npm verb stack WebLoginNotSupported: Web login not supported
...
npm verb statusCode 404
npm verb cwd C:\temp\t1
npm verb Windows_NT 10.0.19044
npm verb node v18.12.1
npm verb npm  v9.1.2
npm ERR! code ENYI
npm ERR! Web login not supported
npm verb exit 1
npm verb code 1
```

It works with --auth-type=legacy, though.

**Comment by lukekarrys** [comment]
Added items in the body of this issue for how to maintain lockfile compatibility with `npm@6` and continue installing local dependencies as symlinks.

**Comment by DevCon1100** [comment]
Access

**Comment by shifenhutu** [comment]
do we have any plan to support `creates hard links from the global store to the project's folders.`  

 like  `pnpm` 

https://pnpm.io/faq#why-does-my-node_modules-folder-use-disk-space-if-packages-are-stored-in-a-global-store

**Comment by ljharb** [comment]
@hutuyishi https://github.com/npm/rfcs/blob/main/accepted/0042-isolated-mode.md already exists for symlinks; i'm not sure if hard links are planned since they add a lot of complexity, especially across platforms.

**Comment by jmcelroy58** [comment]
Is npm login still supported in v9.x?

**Comment by yjhmelody** [comment]
When will `package.json` support comment just like `tsconfig.json`

**Comment by ljharb** [comment]
@yjhmelody that would be a bad idea prior to node supporting serializing and deserializing such a format.

**Comment by rondales** [comment]
> ## 🎉 `npm i npm -g` will now install `npm@9` 🎉
> ### Breaking Changes
> * For a full list of the breaking changes, see [this GitHub Changelog post](https://github.blog/changelog/2022-10-24-npm-v9-0-0-released/) or the initial [`npm@9.0.0`](https://github.com/npm/cli/releases/tag/v9.0.0) release
> 
> ## Migration Guide
> ### Installing symlinked local dependencies
> `npm@9` changed the default value of `install-links` from `false` to `true`.
> 
> This means that npm will now attempt to install directories by default instead of symlinking them.
> 
> Note that workspaces will always be symlinked, so the latest workspaces changes will always be reflected in your package.
> 
> **How to turn it off**
> 
> You can run the following to set the previous behavior on a per-project basis:
> 
> ```
> npm config set install-links=false --location project
> ```
> 
> ### Compatibility with `npm@6`
> `npm@9` changed the `lockfile-version` for new projects from `2` to `3`.
> 
> If you are creating a **new project** and want to maintain compatibility with `npm@6` you will need to downgrade your `lockfile-version` to `2`. Existing projects will not have their `lockfile-version` number changed.
> 
> To set a project specific `lockfile-version` run:
> 
> ```
> npm config set lockfile-version=2 --location project
> ```
> 
> ## Known Issues
> ### `ENOENT` `package.json` Errors
> > **Note**:
> > If you get this error, try running the command with `--install-links=false`
> 
> `npm@9` has changed the default of the `install-links` flag to `true`. This means that when running `npm install` with a directory as the argument (such as `npm install ../path/to/local-package`), `npm` will attempt to install that directory into the `node_modules/` directory just like any other package, instead of creating a symlink like it would previously.
> 
> We are tracking a fix (ref: #5820) for a known issue where transitive local directory dependencies can cause this error.
> 
> **Workaround**
> 
> To workaround this issue, run either of the following:
> 
> * `npm install --install-links=false` to run a single install that will create symlinks
> * `npm config set install-links=false --location project` to set this as a default for your project in its local `.npmrc file`
> 
> ## Other Issues
> If you find other issues with `v9`, use the [Bug v9 Issue Template](https://github.com/npm/cli/issues/new?assignees=&labels=Bug%2CNeeds+Triage%2CRelease+9.x&template=bug_9.yml&title=%5BBUG%5D+%3Ctitle%3E) to open a new issue.

**Comment by tallero** [comment]
I'm nobody to say anything but I frankly still [haven't got how to actually install projects and not symlinks to them](https://github.com/npm/cli/issues/5965#issuecomment-2697481929), despite this FAQs saying actually it should be the default choice.
